Build a file-based catalog

clobrano opened the pull request:
<!-- Thanks for contributing to our project! We appreciate your time and effort.
Please fill out the information below to expedite the review and merge of your pull request.
-->

#### Why we need this PR

Currently NHC's Makefile creates a db-based catalog, which is deprecated

```
DEPRECATION NOTICE:
Sqlite-based catalogs and their related subcommands are deprecated. Support for
them will be removed in a future release. Please migrate your catalog workflows
to the new file-based catalog format.
```

#### Changes made

This change creates a file-based catalog

https://docs.openshift.com/container-platform/latest/operators/admin/olm-managing-custom-catalogs.html#olm-managing-custom-catalogs-fb

**NOTE:** the catalog image changed name `node-healthcheck-operator-index` -> `node-healthcheck-operator-catalog` to be aligned with the other operators.


#### Which issue(s) this PR fixes

https://issues.redhat.com/browse/ECOPROJECT-1948

#### Test plan

- create the file-based catalog and push it to my quay.io repository via Makefile
- create a CatalogSource pointing to the new catalog
- install FAR via WebUI, verify that everything works


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

* **New Features**
  * Added a file-based catalog generation and publishing flow as part of bundle publishing.
* **Changes**
  * Updated the container publish process to build and push the catalog image instead of the previous index image.
  * Catalog channel/version entries (including optional replace/skip-range behavior) are now appended for improved consistency.
  * Bundle icon generation is now driven by a configurable icon asset.
* **CI**
  * Updated the e2e Kubernetes job to use additional render flags (`--skip-tls-verify`).
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

clobrano committed:
Build a file-based catalog

Replace the deprecated db-based catalog with a file-based one.

Signed-off-by: Carlo Lobrano <c.lobrano@gmail.com>

clobrano commented:
interesting, it seems kind isn't happy with this change

```
2024/04/05 17:09:58 render reference "kind-registry:5000/node-healthcheck-operator-bundle:latest": error resolving name for image ref kind-registry:5000/node-healthcheck-operator-bundle:latest: failed to do request: Head "https://kind-registry:5000/v2/node-healthcheck-operator-bundle/manifests/latest": http: server gave HTTP response to HTTPS client
```

clobrano commented:
I am testing the k8s-e2e phase, do not merge
/hold

razo7 commented on Makefile:
```diff
@@ -459,17 +459,45 @@ bundle-cleanup: operator-sdk ## Remove bundle installed via bundle-run
-index-build: opm ## Build a catalog image.
-	$(OPM) index add --container-tool podman --mode semver --tag $(INDEX_IMG) --bundles $(BUNDLE_IMG)
+# Build a file-based catalog image
+# https://docs.openshift.com/container-platform/4.14/operators/admin/olm-managing-custom-catalogs.html#olm-managing-custom-catalogs-fb
```
Please update this line :)

```suggestion
# https://docs.openshift.com/container-platform/4.21/operators/admin/olm-managing-custom-catalogs.html#olm-managing-custom-catalogs-fb
```

razo7 commented on Makefile:
```diff
@@ -459,17 +459,45 @@ bundle-cleanup: operator-sdk ## Remove bundle installed via bundle-run
+	$(OPM) init ${OPERATOR_NAME} \
+		--default-channel=${CHANNELS} \
+		--description=./README.md \
+		--icon=${BLUE_ICON_PATH} \
```
`BLUE_ICON_PATH` is missing, so please add that to avoid https://github.com/medik8s/fence-agents-remediation/pull/169 changes in a follow up

razo7 reviewed (commented):
The FBC migration (Makefile changes) looks correct — it matches what was merged for SNR ([#252](https://github.com/medik8s/self-node-remediation/pull/252)), NMO ([#143]( https://github.com/medik8s/node-maintenance-operator/pull/143)), FAR ([#138](https://github.com/medik8s/fence-agents-remediation/pull/138)), and MDR ([#123](https://github.com/medik8s/machine-deletion-remediation/pull/123)).

I did left small improvements, and please drop the`[test] unsecure registry debug info` commit as it should not be relevant to this PR.

razo7 commented on Makefile:
```diff
@@ -95,9 +95,9 @@ OPERATOR_NAMESPACE ?= openshift-workload-availability
-# You can use it as an arg. (E.g make bundle-build INDEX_IMG=<some-registry>/<project-name-index>:<tag>)
-INDEX_IMG ?= $(IMAGE_REGISTRY)/$(OPERATOR_NAME)-index:$(IMAGE_TAG)
+# The image tag given to the resulting catalog image (e.g. make catalog-build CATALOG_IMG=example.com/operator-catalog:v0.2.0).
+# NOTE: keeping the `-index` appendix to keep the same name already used for the catalog.
+CATALOG_IMG ?= $(IMAGE_REGISTRY)/$(OPERATOR_NAME)-index:$(IMAGE_TAG)
```
We are removing the notion of index in all the of the occurences in the Makefile, so I think we can remove it over here as well. In other operators, we don't call this image an `index`, and we call it a `catalog` (related to File-Based Catalog (FBC)).
```suggestion
CATALOG_IMG ?= $(IMAGE_REGISTRY)/$(OPERATOR_NAME)-catalog:$(IMAGE_TAG)
```
WDYT? @clobrano

clobrano replied:
Sorry for the late response. I agree on that, but I also need to review the state of the repo before moving on, this PR is ancient :sweat_smile:

razo7 replied:
Works for me, and no rush

It is a small PR (and easy to review IMHO), I thought we could finish it quickly :D

clobrano commented:
Checking the pre-build error...

clobrano commented on Makefile:
```diff
@@ -491,7 +520,7 @@ container-build-metrics: ## Build containers for K8s with metric related configu
 .PHONY: container-push
 container-push:  ## Push containers (NOTE: catalog can't be build before bundle was pushed)
-	make docker-push bundle-push index-build index-push
+	make docker-push bundle-push catalog-build catalog-push
```
There is no plan to use `-j`

clobrano commented on Makefile:
```diff
@@ -448,17 +449,45 @@ bundle-cleanup: operator-sdk ## Remove bundle installed via bundle-run
+	$(OPM) generate dockerfile ${CATALOG_DIR}
+	$(OPM) init ${OPERATOR_NAME} \
+		--default-channel=${CHANNELS} \
+		--description=./README.md \
```
Not comma-separated. CHANNELS can have just one of the suggested values

clobrano replied:
You were actually right here

clobrano commented on Makefile:
```diff
@@ -448,17 +449,45 @@ bundle-cleanup: operator-sdk ## Remove bundle installed via bundle-run
+		--icon=${BLUE_ICON_PATH} \
+		--output yaml \
+		> ${CATALOG_INDEX}
+	$(OPM) render ${BUNDLE_IMG} --output yaml >> ${CATALOG_INDEX}
```
Unless specific use cases, I don't think we should allow HTTP or skip TLS

clobrano replied:
On the other end, OPM is failing in e2e-k8s job because we use an insecure kind-registry for the test. I'm considering to set an empty `OPM_RENDER_FLAGS` for all cases, which is set to `--skip-tls-verify` for this test

clobrano committed:
Add configurable OPM_RENDER_FLAGS

The OPM_RENDER_FLAGS defaults to empty.
It is set to `--skip-tls-verify` in pre-submit e2e-k8s NHC_build step to
support the insecure kind registry.

clobrano commented:
Need to fix `opm render` with empty `OPM_RENDER_FLAGS`
/hold

clobrano commented:
`OPM_RENDER_FLAGS` problem fixed.
I also configured the new `quay.io/medik8s/node-healthcheck-operator-catalog` repository

/unhold

razo7 commented on Makefile:
```diff
@@ -25,6 +25,8 @@ YQ_VERSION = v4.53.2
 # ENVTEST_K8S_VERSION refers to the version of kubebuilder assets to be downloaded by envtest binary.
 ENVTEST_K8S_VERSION = 1.34
 
+BLUE_ICON_PATH = "./config/assets/nhc_blue.png"
```
Please use `BLUE_ICON_PATH `in a few other places of Makefile as https://github.com/medik8s/node-healthcheck-operator/blob/main/Makefile#L352C55-L352C63 and  https://github.com/medik8s/node-healthcheck-operator/blob/main/Makefile#L396

clobrano replied:
Ah, sorry about that

razo7 commented on .github/workflows/pre-submit.yaml:
```diff
@@ -40,6 +40,7 @@ jobs:
     timeout-minutes: 30
     env:
       IMAGE_REGISTRY: kind-registry:5000
+      OPM_RENDER_FLAGS: --skip-tls-verify
```
Is this change related to the PR?

razo7 replied:
Can you mention this notion for `OPM_RENDER_FLAGS` in the PR description for easier distinction to future us? @clobrano

slintes reviewed (changes requested):
- multiple channels does not work correctly
- replaces and skip range are not set

CHANNELS=stable,test VERSION=0.4.0 PREVIOUS_VERSION=0.3.1 SKIP_RANGE_LOWER=0.2.0 make catalog-build

slintes commented:
edit: nevermind,  I should not use old version numbers, it pulled the old bundle where we still had that dependency...

~~huh, and didn't we remove that dependency a long while ago? Where is that coming from 🤔~~
```
     - type: olm.gvk.required
       value:
         group: self-node-remediation.medik8s.io
         kind: SelfNodeRemediation
         version: v1alpha1
```

clobrano committed:
fix: replace hardcoded image paths with BLUE_ICON_PATH

clobrano committed:
fix: use appropriate default-image for the catalog

clobrano committed:
fix: create a correct olm.channel blob for each channel in CHANNELS

- create one olm.channel blob for each channel
- add missing replaces field
- add missing skipRange field

clobrano commented:
Failure of 5.0-openshift-e2e seems unlikely related to this change

/retest-required
